lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -197,17 +213,54 @@ void SPIRVRegularizeLLVM::buildFunnelShiftLeftFunc(Function *FSHLFunc) {
   Builder.CreateRet(FunnelShiftRes);
 }
 
-void SPIRVRegularizeLLVM::lowerFunnelShiftLeft(IntrinsicInst *FSHLIntrinsic) {
+void SPIRVRegularizeLLVM::buildFunnelShiftRightFunc(Function *FSHRFunc) {
+  if (!FSHRFunc->empty())
+    return;
+
+  auto *RotateBB = BasicBlock::Create(M->getContext(), "rotate", FSHRFunc);
+  IRBuilder<> Builder(RotateBB);
+  Type *Ty = FSHRFunc->getReturnType();
+  // Build the actual funnel shift rotate logic.
+  // In the comments, "int" is used interchangeably with "vector of int
+  // elements".
+  FixedVectorType *VectorTy = dyn_cast<FixedVectorType>(Ty);
+  Type *IntTy = VectorTy ? VectorTy->getElementType() : Ty;
+  unsigned BitWidth = IntTy->getIntegerBitWidth();
+  ConstantInt *BitWidthConstant = Builder.getInt({BitWidth, BitWidth});
+  Value *BitWidthForInsts =
+      VectorTy ? Builder.CreateVectorSplat(VectorTy->getNumElements(),
+                                           BitWidthConstant)
+               : BitWidthConstant;
+  auto *RotateModVal =
+      Builder.CreateURem(/*Rotate*/ FSHRFunc->getArg(2), BitWidthForInsts);
+  // Shift the less significant number right, the "rotate" number of bits
+  // will be 0-filled on the left as a result of this regular shift.
+  auto *ShiftRight = Builder.CreateLShr(FSHRFunc->getArg(0), RotateModVal);
+  // We want the "rotate" number of the second int's LSBs to occupy the
+  // leftmost "0 space" left by the previous operation. Therefore,
+  // subtract the "rotate" number from the integer bitsize...
+  auto *SubRotateVal = Builder.CreateSub(BitWidthForInsts, RotateModVal);
+  // ...and left-shift the second int by this number, zero-filling the LSBs.
```

Comment:
Similarly to the above:
```suggestion
  // ...and left-shift the more significant int by this number, zero-filling the LSBs.
```